### PR TITLE
feat: Introduce metrics class for streaming consumer

### PR DESCRIPTION
**Pull request by lynnagara**
The purpose of this change is to remove the dependency that currently
exists between streaming consumer code (specifically the stream processor
and parallel processing steps) to Snuba's metrics module.

This is solved by introducing a separate metrics abstract class (this
will eventually move out of Snuba's codebase into the streaming library)
as well as an adapter (this will stay in Snuba) to convert Snuba's metrics
backend to the metrics class expected by the future streaming library. We
also copy over the Gauge implementation as this can no longer depend on the
version in snuba.utils.metrics.

The `metrics` argument no longer needs to be passed to StreamProcessor or
ParallelTransformStep. If this is not present a default (which records nothing)
will be used.

### CODE REVIEW

File: snuba/utils/streams/metrics.py [hunk 1]
```diff
@@ -1,10 +1,11 @@
...
-from typing import Any, Mapping, Optional, Union
+from abc import abstractmethod
+from typing import Any, Mapping, Optional, Protocol, Union, runtime_checkable
 
 Tags = Optional[Mapping[str, str]]
 
 
-class Metrics(ABC):
+@runtime_checkable
+class Metrics(Protocol):
```
Comment (fpacifici) [inline snuba/utils/streams/metrics.py:8]: I guess you turned this into a protocol to be able to depend on this interface without the sentry and Snuba metric classes to extend the same abstract class. Is this right? If yes, great idea.

File: snuba/utils/streams/processing/strategies/streaming/transform.py [hunk 2]
```diff
@@ -280,8 +278,9 @@ def __init__(
         max_batch_time: float,
         input_block_size: int,
         output_block_size: int,
-        metrics: MetricsBackend,
+        metrics: Metrics = DummyMetricsBackend,
     ) -> None:
+        assert isinstance(metrics, Metrics)
```
Comment (fpacifici) [inline snuba/utils/streams/processing/strategies/streaming/transform.py:283]: Why is this needed ?

**Review by fpacifici — COMMENTED** [review]
One clarification otherwise this seems good.

File: snuba/utils/streams/processing/strategies/streaming/transform.py (hunk 2 above)
Comment (lynnagara) [reply snuba/utils/streams/processing/strategies/streaming/transform.py:283]: It's not really needed, but I figured it's nicer to error here if something is wrong with `metrics` than later when .gauge(), .timing() or .increment() are used.

Comment (fpacifici) [reply snuba/utils/streams/processing/strategies/streaming/transform.py:283]: Then could you please add a comment on the reason why the `isinstance` call is there ?

Comment (lynnagara) [reply snuba/utils/streams/processing/strategies/streaming/transform.py:285]: done

**Review by fpacifici — APPROVED** [review]
